diff -Naur old/kernel-2.6.18/xen/arch/x86/domain_build.c new/kernel-2.6.18/xen/arch/x86/domain_build.c
--- old/kernel-2.6.18/xen/arch/x86/domain_build.c	2011-05-17 12:58:03.000000000 -0400
+++ new/kernel-2.6.18/xen/arch/x86/domain_build.c	2011-05-17 13:21:36.000000000 -0400
@@ -345,8 +345,12 @@
     if ( parms.pae == PAEKERN_extended_cr3 )
             set_bit(VMASST_TYPE_pae_extended_cr3, &d->vm_assist);
 
-    if ( UNSET_ADDR != parms.virt_hv_start_low && elf_32bit(&elf) )
-    {
+    // DJWILL: applied Ian Jackson's patch 
+    //      if ( UNSET_ADDR != parms.virt_hv_start_low && elf_32bit(&elf) )
+    //        {
+    if ( elf_32bit(&elf) ){
+        uint64_t virt_hv_start_low;
+
 #if CONFIG_PAGING_LEVELS < 4
         unsigned long mask = (1UL << L2_PAGETABLE_SHIFT) - 1;
 #else
@@ -354,14 +358,22 @@
                              ? (1UL << L2_PAGETABLE_SHIFT) - 1
                              : (1UL << L4_PAGETABLE_SHIFT) - 1;
 #endif
+        // DJWILL: applied Ian Jackson's patch 
+        //value = (parms.virt_hv_start_low + mask) & ~mask;
+        virt_hv_start_low = parms.virt_hv_start_low;
+        if (virt_hv_start_low == UNSET_ADDR)
+            virt_hv_start_low = __HYPERVISOR_COMPAT_VIRT_START;
+        value = (virt_hv_start_low + mask) & ~mask;
 
-        value = (parms.virt_hv_start_low + mask) & ~mask;
 #ifdef CONFIG_COMPAT
         HYPERVISOR_COMPAT_VIRT_START(d) =
             max_t(unsigned int, m2p_compat_vstart, value);
         d->arch.physaddr_bitsize =
             fls((1UL << 32) - HYPERVISOR_COMPAT_VIRT_START(d)) - 1
-            + (PAGE_SIZE - 2);
+            + (PAGE_SHIFT - 2);
+            //            + (PAGE_SIZE - 2);
+        //DJWILL: applied Jan Beulich's patch of PAGE_SIZE->PAGE_SHIFT
+
         if ( value > (!is_pv_32on64_domain(d) ?
                       HYPERVISOR_VIRT_START :
                       __HYPERVISOR_COMPAT_VIRT_START) )
diff -Naur old/kernel-2.6.18/xen/arch/x86/domain.c new/kernel-2.6.18/xen/arch/x86/domain.c
--- old/kernel-2.6.18/xen/arch/x86/domain.c	2011-05-17 12:58:03.000000000 -0400
+++ new/kernel-2.6.18/xen/arch/x86/domain.c	2011-05-17 13:21:36.000000000 -0400
@@ -316,7 +316,9 @@
 
     d->arch.physaddr_bitsize =
         fls((1UL << 32) - HYPERVISOR_COMPAT_VIRT_START(d)) - 1
-        + (PAGE_SIZE - 2);
+            + (PAGE_SHIFT - 2);
+        //            + (PAGE_SIZE - 2);
+        //DJWILL: applied Jan Beulich's patch of PAGE_SIZE->PAGE_SHIFT
 
     return 0;
 
